libraries/lib-numeric-formats/formatters/BeatsNumericConverterFormatter.cpp

```diff
@@ -0,0 +1,359 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/**********************************************************************
+
+ Audacity: A Digital Audio Editor
+
+ @file BeatsNumericConverterFormatter.cpp
+
+ Dmitry Vedenko
+
+ **********************************************************************/
+#include "BeatsNumericConverterFormatter.h"
+
+#include <algorithm>
+#include <array>
+#include <cmath>
+
+#include "NumericConverterRegistry.h"
+#include "NumericConverterFormatterContext.h"
+
+#include "SampleCount.h"
+
+#include "Project.h"
+#include "ProjectTimeSignature.h"
+
+namespace
+{
+// This function will return 10^pow
+// No overflow checks are performed, it is assumed that 10^pow
+// does not overflow
+constexpr size_t Get10Pow (size_t pow)
+{
+   return pow > 0 ? 10 * Get10Pow(pow - 1) : 1;
+}
+
+/* i18-hint: The music theory "bar" */
+const auto BarString = XO("bar");
+/* i18-hint: The music theory "beat" */
+const auto BeatString = XO("beat");
+
+class BeatsFormatter final :
+    public NumericConverterFormatter,
+    public PrefsListener
+{
+public:
+   static constexpr std::array<size_t, 3> MIN_DIGITS { 3, 2, 2 };
+   static constexpr std::array<size_t, 3> UPPER_BOUNDS {
+      Get10Pow(MIN_DIGITS[0] - 1) + 1, Get10Pow(MIN_DIGITS[1] - 1) + 1,
+      Get10Pow(MIN_DIGITS[2] - 1) + 1
+   };
+   
+   BeatsFormatter(const FormatterContext& context, int fracPart)
+       : mContext { context }
+       , mFracPart { fracPart }
+   {
+      auto project = mContext.GetProject();
+
+      if (!project)
+         return;
+
+      mBarString = BarString.Translation();
+      mBeatString = BeatString.Translation();
+
+      UpdateFormat(*project);
+
+      // Subscribing requires non-const reference
+      mTimeSignatureChangedSubscription =
+         const_cast<ProjectTimeSignature&>(ProjectTimeSignature::Get(*project))
+            .Subscribe(
+               [this](const auto&)
+               {
+                  // Receiving this message means that project is
+                  // alive and well
+                  UpdateFormat(*mContext.GetProject());
+                  Publish({});
+               });
+   }
+
+   bool CheckField(size_t fieldIndex, int value) const noexcept
+   {
+      if (fieldIndex >= mFields.size())
+         return false;
+
+      const auto digitsCount = mFields[fieldIndex].digits;
+
+      // Format always allows at least two digits
+      const auto lowerRange =
+         digitsCount > MIN_DIGITS[fieldIndex] ? Get10Pow(digitsCount - 1) : 0;
+
+      const auto upperRange = Get10Pow(digitsCount);
+
+      return value >= int(lowerRange) && value < int(upperRange);
```

Comment:
This function checks that the width allocated to the field is at least the minimum number of digits, and otherwise just sufficient to show the given, maximum, 1-based value.

I would comment that.